sdk/identity/Azure.Identity/src/SharedTokenCacheCredential.cs

```diff
@@ -128,53 +138,46 @@ private async ValueTask<ExtendedAccessToken> GetTokenImplAsync(TokenRequestConte
 
         private async Task<(IAccount, Exception)> GetAccountAsync()
         {
-            Exception ex = null;
+            List<IAccount> accounts = (await _client.GetAccountsAsync().ConfigureAwait(false)).ToList();
 
-            IAccount account = null;
+            List<IAccount> filteredAccounts = accounts.Where(a => (string.IsNullOrEmpty(_username) || a.Username == _username) && (string.IsNullOrEmpty(_tenantId) || a.HomeAccountId?.TenantId == _tenantId)).ToList();
+
+            if (filteredAccounts.Count != 1)
+            {
+                return (null, new CredentialUnavailableException(GetCredentialUnavailableMessage(accounts, filteredAccounts)));
+            }
 
-            List<IAccount> allAccounts = (await _client.GetAccountsAsync().ConfigureAwait(false)).ToList();
+            return (filteredAccounts.First(), null);
+        }
 
-            List<IAccount> accounts = (!string.IsNullOrEmpty(_username)) ? allAccounts.Where(a => a.Username == _username).ToList() : allAccounts.ToList();
+        internal const string NoAccountsInCacheMessage = "No accounts were found in the cache.  To authenticate with the SharedTokenCacheCredential, login an account through developer tooling supporting Azure single sign on.";
+        internal const string MultipleAccountsInCacheMessage = "Multiple accounts were found in the cache. To authenticate with the SharedTokenCacheCredential, set the AZURE_USERNAME and AZURE_TENANT_ID environment variables to the preferred username and tenantId, or specify them to the constructor. {0}";
```

Comment:
Where do we read `AZURE_TENANT_ID` env var for shared token credential?

Reply:
Found it

Reply:
In `DefaultAzureCredential`. Currently both `AZURE_USERNAME` and `AZURE_TENANT_ID` are only respected when instantiating the `SharedTokenCacheCredential` through the `DefaultAzureCredential`. I'm on the fence as to whether to change this behavior. Currently only the `DefaultAzureCredential` and the `EnvironmentCredential` actually read environment variables and the rest of the credential types need to be explicitly passed values for these. It's possible to change this so the `SharedTokenCacheCredential` will read the environment variable itself, but then perhaps all other credential types such as `ClientSecretCredential`, `ClientCertificateCredential`, `InteractiveBrowserCredential` etc. would also directly respect these environment variables.

I'm not opposed to changing this behavior throughout the library, but if we do, it should be done as part of a larger work item.

Reply:
File an issue to decide this please. We need to do it soon/

Reply:
Added issue https://github.com/Azure/azure-sdk-for-net/issues/8692 to track this.